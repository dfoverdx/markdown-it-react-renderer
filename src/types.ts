import { ReactNode } from 'react';
import MarkdownIt from 'markdown-it';
import { Options } from 'markdown-it/lib/index';
import MDIToken from 'markdown-it/lib/token';

declare class ReactRenderer {
    rules: {
        [key: string]: TokenRender;
    };

    render(tokens: Token[], options: any, env: any): ReactNode;
};

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

export type MarkdownItPresetName = 'commonmark' | 'zero' | 'default';

/**
 * A MarkdownIt Token with the overloaded capacity to allow the `content` property to be a `ReactNode`.
 */
export interface Token extends Omit<MDIToken, 'content'> {
    content: ReactNode;
}

/**
 * Rule defintion that modifies the tokens of a `Token` stream.
 * 
 * @see https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md#rules
 */
export type TokenRender = (tokens: Token[], idx: number, options: any, env: any, self: ReactRenderer) => 
    ReactNode | RenderedToken;

export class RenderedToken {
    constructor(public node: ReactNode, public endIdx: number) {}
}

/**
 * Key-value pairs of `TokenRender` rules.
 */
export interface MarkdownItRules {
    [key: string]: TokenRender;
}

export type Plugin = [(md: MarkdownIt, env: any) => void, ...any[]] | ((md: MarkdownIt, env: any) => void);

export interface ReactifyMarkdownProps<T extends 'md' | 'options'> {
    /**
     * The markdown source string to be parsed and rendered.
     */
    children: string;

    /**
     * Environment context.  Gets passed when evaluating rules, which may be helpful for rendering React components.
     */
    env: any;

    /**
     * By default, the component will strip preceding spaces via `strip-indent`.  Set this flag to disable this 
     * behavior.
     * 
     * @see https://github.com/sindresorhus/strip-indent
     */
    dontStripIndent: boolean;

    /**
     * The pre-configured `MarkdownIt` instance.
     */
    md: T extends 'md' ? MarkdownIt : never;

    /**
     * Optional rule name or list of rule names to enable.
     */
    enable?: T extends 'options' ? (string | string[]) : never;
    
    /**
     * Optional rule name or list of rule names to disable.
     */
    disable?: T extends 'options' ? (string | string[]) : never;

        /**
     * Options passed to the `MarkdownIt` constructor.
     * 
     * @see https://markdown-it.github.io/markdown-it/#MarkdownIt.new
     */
    options?: T extends 'options' ? Options : never;

    /**
     * List of plugins to apply to the `MarkdownIt` instance.
     * 
     * @warning Any plugin which overwrites the instance's renderer directly will cause an error.
     */
    plugins?: T extends 'options' ? (Plugin[] | Plugin) : never;

    /**
     * The preset name to use.
     * 
     * @default `"default"`
     * @see https://markdown-it.github.io/markdown-it/#MarkdownIt.new
     */
    presetName?: T extends 'options' ? MarkdownItPresetName : never;
}