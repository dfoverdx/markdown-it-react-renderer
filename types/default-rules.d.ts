import { MarkdownItRules } from './types';
/**
 * Reactified rules defined in markdown-it/lib/renderer.js.  May need to update these with future versions of
 * markdown-it.
 */
declare const default_rules: MarkdownItRules;
export default default_rules;
