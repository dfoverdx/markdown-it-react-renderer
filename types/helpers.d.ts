import { Token } from './types';
/**
 * Converts a `Token`'s attribute tuples into an object.
 *
 * @param token The `Token` from which to retrieve attributes.
 */
export declare function getAttrs(token: Token): {
    [key: string]: string;
};
