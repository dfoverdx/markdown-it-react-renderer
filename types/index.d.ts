import MarkdownIt from 'markdown-it';
import './types';
export { default as ReactRenderer } from './react-renderer';
export * from './types';
export { default } from './reactify-markdown';
/**
 * Plugin that replaces the MarkdownIt instance's `renderer` with one that renders to `ReactNode`s.
 * @param md MarkdownIt instance.
 */
export declare function RenderReactPlugin(md: MarkdownIt): void;
