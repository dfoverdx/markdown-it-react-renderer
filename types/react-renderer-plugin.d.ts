import MarkdownIt from 'markdown-it';
/**
 * Plugin that replaces the MarkdownIt instance's `renderer` with one that renders to `ReactNode`s.
 * @param md MarkdownIt instance.
 */
export default function RenderReactPlugin(md: MarkdownIt): void;
