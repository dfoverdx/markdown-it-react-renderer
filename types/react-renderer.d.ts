import { ReactNode } from 'react';
import MDIToken from 'markdown-it/lib/token';
import { Token, TokenRender } from './types';
/**
 * Renderer that renders the converted source as React elements rather than an HTML string.
 */
export default class ReactRenderer {
    constructor();
    /**
     * Rules to use when parsing and rendering a source string.
     */
    rules: {
        [key: string]: TokenRender;
    };
    /**
     * Renders the passed token stream as React elements.
     *
     * @param tokens Token stream to be rendered as React elements.
     * @param options Options passed to Renderer rules.
     * @param env Environment passed to Renderee rules.
     */
    render(tokens: Token[], options: any, env: any): ReactNode;
    /**
     * Renders the passed MarkdownIt tokens.
     *
     * @deprecated Use JSX Rendered tokens instead, as their `content` property can store React components.
     */
    render(tokens: MDIToken[], options: any, env: any): string;
    /**
     * Does the woek of rendering.  Is in a separate function from `render` so that we can include the `idx` argument.
     *
     * @param tokens The token stream produced by the parser.
     * @param keyPrefix A prefix for element keys.
     * @param idx The index of the starting token to be rendered.  Is `0` when called by `render`, and greater when
     * being called recursively.
     * @param options Options passed to the Renderer instance.
     * @param env Environement passed to the Renderer instance.
     */
    private renderInner;
}
