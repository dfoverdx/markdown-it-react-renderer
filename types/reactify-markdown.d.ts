import { PureComponent } from 'react';
import MD from 'markdown-it';
import { MarkdownItPresetName, ReactifyMarkdownProps as Props } from './types';
interface State {
    md: MD;
}
export default class ReactifyMarkdown<TProps extends Props<'md'> | Props<'options'>> extends PureComponent<TProps, State> {
    constructor(props: TProps);
    componentDidUpdate(nextProps: TProps, nextState: State): void;
    private generateState;
    render(): string;
    static defaultProps: Partial<Props<'options'> | Omit<Props<'md'>, 'presetName'>> & {
        presetName: MarkdownItPresetName;
    };
}
declare type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
export {};
