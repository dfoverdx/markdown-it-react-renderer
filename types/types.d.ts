import { ReactNode } from 'react';
import MarkdownIt from 'markdown-it';
import MDIToken from 'markdown-it/lib/token';
import ReactRenderer from './react-renderer';
declare type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;
/**
 * A MarkdownIt Token with the overloaded capacity to allow the `content` property to be a `ReactNkde`.
 */
export interface Token extends Omit<MDIToken, 'content'> {
    content: ReactNode;
}
/**
 * Rule defintion that modifies the tokens of a `Token` stream.
 *
 * @see https://github.com/markdown-it/markdown-it/blob/master/docs/architecture.md#rules
 */
export declare type TokenRender = (tokens: Token[], idx: number, options: any, env: any, self: ReactRenderer) => ReactNode;
/**
 * Key-value pairs of `TokenRender` rules.
 */
export interface MarkdownItRules {
    [key: string]: TokenRender;
}
export declare type Plugin = [(md: MarkdownIt, env: any) => void, ...any[]];
export {};
