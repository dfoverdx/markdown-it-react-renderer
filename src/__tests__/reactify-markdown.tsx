import React from 'react';
import { shallow } from 'enzyme';
import mdiRegex from 'markdown-it-regex';
import ReactifyMarkdown from '../reactify-markdown';
import { Plugin } from '../types';

jest.mock('shortid');

test('It handles some basic rules', () => {
    let component = shallow(<ReactifyMarkdown>{`
        Header 1
        ========

        A paragraph starts here.  *This is italicized.*

            This is a code block.
    `}</ReactifyMarkdown>);

    expect(component).toMatchSnapshot();
});

test('It handles custom rules', () => {
    let plugins: Plugin[] = [[mdiRegex, {
            name: 'test',
            regex: /:(\w+):/,
            replace: (m: string) => <i>{m}</i>,
        }]];

    let component = shallow(<ReactifyMarkdown plugins={plugins}>{`
        Hay :needle: stack
    `}</ReactifyMarkdown>);

    expect(component).toMatchSnapshot();
})

test('It doesn\'t strip whitespace when it\'s set not to', () => {
    const md = `
    Header 1
    ========

    A paragraph starts here.  *This is italicized.*
    `;

    const component = shallow(<ReactifyMarkdown dontStripIndent>{md}</ReactifyMarkdown>);
    expect(component).toMatchSnapshot();
});