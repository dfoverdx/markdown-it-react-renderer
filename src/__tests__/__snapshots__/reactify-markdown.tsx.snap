// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't strip whitespace when it's set not to 1`] = `
<pre
  key="0"
>
  <code>
    Header 1
========

A paragraph starts here.  *This is italicized.*

  </code>
</pre>
`;

exports[`handles custom rules 1`] = `
<p
  key="0"
>
  Hay 
  <i
    key="1"
  >
    needle
  </i>
   stack
</p>
`;

exports[`handles passing an instance of markdown-it to props 1`] = `
<p
  key="0"
>
  Hay 
  <i
    key="1"
  >
    needle
  </i>
   stack
</p>
`;

exports[`handles some basic rules 1`] = `
Array [
  <h1
    key="0"
  >
    Header 1
  </h1>,
  <p
    key="3"
  >
    A paragraph starts here.  
    <em
      key="1"
    >
      This is italicized.
    </em>
  </p>,
  <pre
    key="2"
  >
    <code>
      This is a code block.

    </code>
  </pre>,
]
`;

exports[`properly handles new-lines 1`] = `
Array [
  <p
    key="0"
  >
    This is a paragraph.
    

    It does not have a newline.
  </p>,
  <p
    key="3"
  >
    This is a paragraph.
    <br />
    

    It does have a newline.
  </p>,
]
`;

exports[`uses props specified by overwriting ReactifyMarkdown.defaultProps 1`] = `
<p>
  Hay 
  <i>
    needle
  </i>
   stack
</p>
`;

exports[`uses props specified by overwriting ReactifyMarkdown.defaultProps 2`] = `
<p>
  Hay 
  <em>
    needle
  </em>
   stack
</p>
`;

exports[`uses props specified by overwriting ReactifyMarkdown.defaultProps 3`] = `
<p>
  Hay 
  <em>
    needle
  </em>
   stack
</p>
`;

exports[`uses props specified by overwriting ReactifyMarkdown.defaultProps 4`] = `
<p>
  Hay 
  <em>
    needle
  </em>
   stack
</p>
`;
